### PR TITLE
Make bfs_successors and bfs_predecessors return iterators

### PR DESCRIPTION
**Pull request by MridulS**

### CODE REVIEW

**Comment by dschult** [comment]
bfs_successors looks like you create the dict in the function and then yield from it. Maybe an iterator is not a good idea here. Since the dict already exists we might as well return it. What do you think?

**Comment by MridulS** [comment]
I think `bfs_successors` and `bfs_predecessors` for uniformity should return same type of result (iterator/dictionary). If we go with returning dictionary for `bfs_successors` we should also revert `bfs_predecessors` back to dictionary. I think we should go with iterators for 2.0

**Comment by dschult** [comment]
Hmmm...  iterators are used in order to save memory. So makin a dict in order to then provide an iterator sort of defeats the purpose.  But let's try to make it an iterator, as you say--for consistency.

Is there a way to make bfs_successors an iterator that does not create the dict. Take a look at the bfs_edges... A BFS ought to get all the successors of a node before moving to the next node. Can you form the list of successors and yield that before moving to the next node?  Then we'd have a true iterator.

**Comment by MridulS** [comment]
`bfs_edges` returns `parent, child`. So, we can return `bfs_predecessors` as iterators and calling dict(iterator) will club them as {child: parent}, every node will have only one parent.  But that is not the case with  `bfs_successors`, a node can have multiple children. If we make `bfs_successors` an iterator the dict(iterator) version for the user will get complicated because it is non-trivial to produce a dictionary with a list of children as values.  (You have to first use defaultdict(list) and so on). If we move on from the dictionary version to list version then we can use iterators. What do you think?

**Comment by dschult** [comment]
Could the iterator yield ```(node, child_list)``` ?
Then ```dict(bfs_succ(G))``` would produce what the current ```bfs_successors``` code returns.

To get the list of children, we'd have to construct it from many (node, child) values from bfs_edges. Would something like this work?

    parent=None
    children=[]
    for (p,c) in bfs_edges: 
        if p == parent:
            children.append(c)
            continue
        yield (parent, children)
        children = [c]
        parent = p
    yield (parent, children)

**Comment by MridulS** [comment]
@dschult Thanks :)

**Comment by dschult** [comment]
#1826 might be trickier because the edges in a depth first search don't exhaust all the edges from a parent before moving to the next parent.  With Depth First you might have to wait a long time before getting all the successors for a particular node. I'll take a look at what you've got there.

For this PR, your updates make me realize that for the initial case, setting ```parent = None``` might not be the best choice. Is it true that the parent of the first edge is always ```source```?  If so we could initialize parent to ```source``` and remove the check for parent being None.

Thanks for this!!

**Comment by MridulS** [comment]
@dschult Sorry for the absence.
Yes, the parent of the first edge is always `source` as we start `bfs_edges` with `parent, children = queue[0]` where `queue = deque([(source, neighbors(source))])`. So we can start with `parent = source` and remove the check condition. 
I have updated the PR.

Thanks! :)

File: networkx/algorithms/traversal/breadth_first_search.py [hunk 1]
```diff
@@ -108,14 +109,14 @@ def bfs_predecessors(G, source):
 
     Returns
     -------
-    pred: dict
-       A dictionary with nodes as keys and predecessor nodes as values.
+    pred: iterator
+        (key, predecessor) iterator
```
Comment (dschult) [inline networkx/algorithms/traversal/breadth_first_search.py]: should the word "key" here (in doc_string) be "node"? 
Doesn't really matter but perhaps more clear? Same in other doc_strings that use "key"...

**Comment by MridulS** [comment]
@dschult

**Pull request merged**